File: packages/validator-go/validator/expr/token.go
// Package expr is the Go CRUDUI expression engine (EXPRESSION-GRAMMAR §0-§7).
//
// Pipeline: string -[Lexer]-> []Token -[Parser]-> Node (AST) -[Evaluator + data]-> value.
// The three stages are separate (GRAMMAR §0): the lexer knows no grammar, the
// parser knows no data, the evaluator knows no string. No eval, no regex/string
// split evaluation (GRAMMAR §10) — a single forward scan plus recursive descent.
//
// This is the CRUDUI engine. It runs in parallel with the legacy condition_parser in
// package validator (R7); nothing here imports legacy and legacy imports nothing here.
//
// The single truth is the shared 4-language fixture
// tests/fixtures/expr/cases.json (already passing in JS and PHP). Tokens, AST
// shape, and evaluated value/truthy are byte-compatible with the JS reference
// (validator-js ConditionParser.ts + PathResolver.ts). AST serialization omits
// position (Rust holds none → the 4-language equivalence compares position-free
// nodes).
package expr

// TokenType tags a lexical token (EXPRESSION-GRAMMAR §1, JS TokenType parity).
//
// String-valued so a fixture's token.type string ("DOT"|"IDENTIFIER"|...) maps
// directly. WHITESPACE never reaches the token slice (the lexer drops it); EOF
// is always the final token. INVALID is one unrecognized character.
//
// Intentionally absent (GRAMMAR §1·§10): arithmetic + - * / %, function calls,
// method calls, regex, assignment =, bitwise, root path /. Do not add them.
type TokenType string

// The Token* constants enumerate every TokenType the lexer emits
// (EXPRESSION-GRAMMAR §1). Their string values are the canonical fixture token
// type names and are byte-compatible with the JS reference. TokenWhitespace and
// TokenInvalid are lexer-internal: WHITESPACE is dropped before the token slice
// and INVALID marks one unrecognized character.
const (
	TokenString     TokenType = "STRING"
	TokenNumber     TokenType = "NUMBER"
	TokenBoolean    TokenType = "BOOLEAN"
	TokenNull       TokenType = "NULL"
	TokenIdentifier TokenType = "IDENTIFIER"
	TokenDot        TokenType = "DOT"
	TokenDotDot     TokenType = "DOT_DOT"
	TokenAsterisk   TokenType = "ASTERISK"
	TokenEQ         TokenType = "EQ"
	TokenNE         TokenType = "NE"
	TokenGT         TokenType = "GT"
	TokenGE         TokenType = "GE"
	TokenLT         TokenType = "LT"
	TokenLE         TokenType = "LE"
	TokenAnd        TokenType = "AND"
	TokenOr         TokenType = "OR"
	TokenNot        TokenType = "NOT"
	TokenIn         TokenType = "IN"
	TokenNotIn      TokenType = "NOT_IN"
	TokenLParen     TokenType = "LPAREN"
	TokenRParen     TokenType = "RPAREN"
	TokenLBracket   TokenType = "LBRACKET"
	TokenRBracket   TokenType = "RBRACKET"
	TokenComma      TokenType = "COMMA"
	TokenQuestion   TokenType = "QUESTION"
	TokenColon      TokenType = "COLON"
	TokenEOF        TokenType = "EOF"
	TokenWhitespace TokenType = "WHITESPACE"
	TokenInvalid    TokenType = "INVALID"
)

// Token is a single lexical token. Value is the raw source text; Literal is the
// parsed value for value tokens (string content, number int/float, bool, the
// dot count for DOT_DOT) and nil otherwise.
//
// No position field: the 4-language token contract excludes positions (the Rust
// AST holds none), matching the shared fixture's token shape.
type Token struct {
	Type    TokenType
	Value   string
	Literal any
}

// ToMap serializes to the shared-fixture token shape {type, value, literal}.
func (t Token) ToMap() map[string]any {
	return map[string]any{
		"type":    string(t.Type),
		"value":   t.Value,
		"literal": t.Literal,
	}
}

File: packages/validator-rust/src/expr/token.rs
//! CRUDUI expression tokens (EXPRESSION-GRAMMAR §1, JS `types.ts` TokenType parity).
//!
//! `TokenType` is the terminal tag; `Token` is `{type,value,literal}`. WHITESPACE
//! never reaches the token array (the lexer drops it); EOF is always the final
//! token. The serialized shape (`to_value`) byte-matches the shared fixture's
//! `tokens` entries — no `position` field, since the 4-language token contract
//! excludes positions (the Rust AST holds none either).
//!
//! Intentionally absent (GRAMMAR §1·§10): arithmetic `+ - * / %`, function calls,
//! method calls, regex, assignment `=`, bitwise, root path `/`. Do not add them.

use serde_json::{json, Value};

/// Terminal token kind. The string spelling matches the shared fixture's
/// `token.type` ("DOT"|"IDENTIFIER"|…).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenType {
    /// Quoted string literal `'…'` / `"…"`.
    String,
    /// Integer or decimal number literal.
    Number,
    /// `true` / `false`.
    Boolean,
    /// `null`.
    Null,
    /// Path segment / unquoted identifier `[A-Za-z_][A-Za-z0-9_]*`.
    Identifier,
    /// Single leading `.`.
    Dot,
    /// Two-or-more dots `..` / `…` (parent ascent).
    DotDot,
    /// Wildcard `*`.
    Asterisk,
    /// `==`.
    Eq,
    /// `!=`.
    Ne,
    /// `>`.
    Gt,
    /// `>=`.
    Ge,
    /// `<`.
    Lt,
    /// `<=`.
    Le,
    /// `&&`.
    And,
    /// `||`.
    Or,
    /// `!`.
    Not,
    /// `in`.
    In,
    /// `not in`.
    NotIn,
    /// `(`.
    LParen,
    /// `)`.
    RParen,
    /// `[`.
    LBracket,
    /// `]`.
    RBracket,
    /// `,`.
    Comma,
    /// `?`.
    Question,
    /// `:`.
    Colon,
    /// End of input.
    Eof,
    /// Whitespace run (dropped before the token array is built).
    Whitespace,
    /// One unrecognized character.
    Invalid,
}

impl TokenType {
    /// The fixture string tag for this token type.
    pub fn as_str(self) -> &'static str {
        match self {
            TokenType::String => "STRING",
            TokenType::Number => "NUMBER",
            TokenType::Boolean => "BOOLEAN",
            TokenType::Null => "NULL",
            TokenType::Identifier => "IDENTIFIER",
            TokenType::Dot => "DOT",
            TokenType::DotDot => "DOT_DOT",
            TokenType::Asterisk => "ASTERISK",
            TokenType::Eq => "EQ",
            TokenType::Ne => "NE",
            TokenType::Gt => "GT",
            TokenType::Ge => "GE",
            TokenType::Lt => "LT",
            TokenType::Le => "LE",
            TokenType::And => "AND",
            TokenType::Or => "OR",
            TokenType::Not => "NOT",
            TokenType::In => "IN",
            TokenType::NotIn => "NOT_IN",
            TokenType::LParen => "LPAREN",
            TokenType::RParen => "RPAREN",
            TokenType::LBracket => "LBRACKET",
            TokenType::RBracket => "RBRACKET",
            TokenType::Comma => "COMMA",
            TokenType::Question => "QUESTION",
            TokenType::Colon => "COLON",
            TokenType::Eof => "EOF",
            TokenType::Whitespace => "WHITESPACE",
            TokenType::Invalid => "INVALID",
        }
    }
}

/// A literal payload carried by a value token. `value` for STRING/NUMBER/
/// BOOLEAN/NULL, the dot count for DOT_DOT, the identifier text for IDENTIFIER;
/// `None` for everything else.
#[derive(Debug, Clone, PartialEq)]
pub enum Literal {
    /// Decoded string content (STRING) or identifier text (IDENTIFIER).
    Str(String),
    /// Integer number literal (the fixture keeps integers as integers).
    Int(i64),
    /// Decimal number literal.
    Float(f64),
    /// Boolean keyword literal.
    Bool(bool),
    /// Dot count for a DOT_DOT token.
    DotCount(i64),
    /// No literal payload.
    None,
}

impl Literal {
    fn to_value(&self) -> Value {
        match self {
            Literal::Str(s) => Value::String(s.clone()),
            Literal::Int(n) => json!(n),
            Literal::Float(f) => json!(f),
            Literal::Bool(b) => Value::Bool(*b),
            Literal::DotCount(n) => json!(n),
            Literal::None => Value::Null,
        }
    }
}

/// A single lexical token. `value` is the raw source text; `literal` is the
/// parsed payload (or `None`). No position field — the 4-language token contract
/// excludes positions, matching the shared fixture's token shape.
#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    /// The token kind.
    pub token_type: TokenType,
    /// Raw source text for this token.
    pub value: String,
    /// Parsed literal payload, if any.
    pub literal: Literal,
}

impl Token {
    /// Build a token from its type, raw text, and literal payload.
    pub fn new(token_type: TokenType, value: impl Into<String>, literal: Literal) -> Self {
        Token {
            token_type,
            value: value.into(),
            literal,
        }
    }

    /// Serialize to the shared-fixture token shape `{type,value,literal}`.
    pub fn to_value(&self) -> Value {
        json!({
            "type": self.token_type.as_str(),
            "value": self.value,
            "literal": self.literal.to_value(),
        })
    }
}
